src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php

```diff
@@ -55,40 +64,70 @@ public function receive(callable $handler): void
             }
 
             try {
-                $handler($this->serializer->decode([
-                    'body' => $AMQPEnvelope->getBody(),
-                    'headers' => $AMQPEnvelope->getHeaders(),
-                ]));
-
-                $this->connection->ack($AMQPEnvelope);
-            } catch (RejectMessageExceptionInterface $e) {
-                try {
-                    $this->connection->reject($AMQPEnvelope);
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
+                $envelope = $this->serializer->decode([
+                    'body' => $this->currentAmqpEnvelope->getBody(),
+                    'headers' => $this->currentAmqpEnvelope->getHeaders(),
+                ]);
+            } catch (MessageDecodingFailedException $exception) {
+                // invalid message of some type
+                $this->rejectCurrentMessage();
 
-                throw $e;
-            } catch (\AMQPException $e) {
-                throw new TransportException($e->getMessage(), 0, $e);
-            } catch (\Throwable $e) {
-                try {
-                    $this->connection->nack($AMQPEnvelope, AMQP_REQUEUE);
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
+                throw $exception;
+            }
 
-                throw $e;
-            } finally {
-                if (\function_exists('pcntl_signal_dispatch')) {
-                    pcntl_signal_dispatch();
-                }
+            $handler(
+                $envelope->with(new RetryCountStamp($this->connection->getRetryCount($this->currentAmqpEnvelope)))
+            );
+
+            if (\function_exists('pcntl_signal_dispatch')) {
+                pcntl_signal_dispatch();
```

Comment:
This is done by the `Worker` as well. Do we really need it here actually? :)

Reply:
The `Worker` does not do this currently, but, should it perhaps? I'm all for moving more boilerplate to the `Worker` so that transports don't need to bother with it.

Reply:
:+1: for having this in the worker

Reply:
Moved into the Worker

Reply:
You can remove it from here now then :)